Enhance extension documentation (#11801)
gardener/gardener: docs/extensions/registration.md

---
title: Registering Extension Controllers
---

# Registering Extension Controllers

Before Gardener can manage shoot clusters, it needs to know which required and optional extensions are available in the landscape.
The following sections explain the general registration process for extensions.

## `Extension`s

The registration starts by creating [`Extension`](../../example/operator/15-extension.yaml) resources in the garden runtime cluster. 
They represent the single source for all deployment aspects an extension may offer (garden, seed, shoot, admission).

> The `gardener-operator` takes these resources to deploy the extension to the garden runtime cluster, as well as creating corresponding [`ControllerRegistration`](../../example/25-controllerregistration.yaml) and [`ControllerDeployment`](../../example/25-controllerdeployment.yaml) resources in the virtual garden cluster.

Reviewer: I assume by`deploy the extension` is meant the extension controller?

Please see the following example of an `Extension` resource which mainly configures:
- The resource kinds and types the extension is responsible for

> - A Reference the OCI Helm chart for the extension

Reviewer: ```suggestion
- A Reference the OCI Helm chart for the extension controller
```

> - Values for the extension in the garden runtime cluster

Reviewer: I assume those values are optional right?
```suggestion
- Optional values for the helm deployment in the garden runtime cluster
```

> - Values for the deployment in the seed clusters

Reviewer: ```suggestion
- Optional values for the helm deployment in the seed clusters
```

> - A Reference the OCI Helm chart(s) for the extension admission

Reviewer: ```suggestion
- A Reference the OCI Helm chart(s) for the extension admission and optional values
```

> ```yaml

Reviewer: Can you please move values fields below chart related fields as they seem to be optional? Also after that can you align the order of the bullets that explain the example to match the example field order?

* `Always`: Gardener will demand the deployment of the extension controller to seed clusters independent of whether it is actually required or not. This might be helpful if you want to add a new component/controller to all seed clusters by default. Another use-case is to minimize the durations until extension controllers get deployed and ready in case you have highly fluctuating seed clusters.
* `AlwaysExceptNoShoots`: Similar to `Always`, but if the seed does not have any shoots, then the extension is not being deployed. It will be deleted from a seed after the last shoot has been removed from it.

Also, the `.spec.extension.seedSelector` allows to specify a label selector for seed clusters.
Only if it matches the labels of a seed, then it will be deployed to it.
Please note that a seed selector can only be specified for secondary controllers (`primary=false` for all `.spec.resources[]`).

### `Extension` Resource Configurations

> The `Extension` resource allows injecting arbitrary steps into the garden, seed and shoot reconciliation flow that are unknown to Gardener.

Reviewer: ```suggestion
The extensibility contract allows injecting arbitrary steps into the garden, seed and shoot reconciliation flow that are unknown to Gardener.
```

> Hence, it is slightly special and allows further configuration when registering it:

Reviewer: ```suggestion
Hence we can configure at what point in the lifecycle those steps should happen
```
